x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSlice.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvSlice::new, field, start, end);
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException("functions do not support scripting");
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field, start, end);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        MvSlice other = (MvSlice) obj;
+        return Objects.equals(other.field, field) && Objects.equals(other.start, start) && Objects.equals(other.end, end);
+    }
+
+    static int calculateOffset(int oldOffset, int fieldValueCount) {
+        return oldOffset < 0 ? oldOffset + fieldValueCount : oldOffset;
+    }
+
+    @Evaluator(extraName = "Int", mvFunction = "true")
+    static void process(IntBlock.Builder builder, int posistion, IntBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(posistion);
+        int first = field.getFirstValueIndex(posistion);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) builder.appendNull();
+        else {
+            for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
+                builder.appendInt(field.getInt(i));
+            }
+        }
+    }
+
+    @Evaluator(extraName = "Boolean", mvFunction = "true")
+    static void process(BooleanBlock.Builder builder, int posistion, BooleanBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(posistion);
+        int first = field.getFirstValueIndex(posistion);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) builder.appendNull();
+        else {
+            for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
+                builder.appendBoolean(field.getBoolean(i));
+            }
+        }
+    }
+
+    @Evaluator(extraName = "Long", mvFunction = "true")
+    static void process(LongBlock.Builder builder, int posistion, LongBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(posistion);
+        int first = field.getFirstValueIndex(posistion);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) builder.appendNull();
+        else {
+            for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
```

Comment:
Two things!
1. The word "offset" implied to me "offset into the block" not "offset into the array of values". I have no idea what word to use, but I was confused.
2. I'd probably pre-calculate the end-offset - just pulling out the `Math.min() +  first` stuff into a variable.